interfaces/prompting/requestprompts/requestprompts.go

```diff
@@ -0,0 +1,373 @@
+	pdb.mutex.Lock()
+	defer pdb.mutex.Unlock()
+	userEntry, ok := pdb.perUser[user]
+	if !ok || len(userEntry.ByID) == 0 {
+		return nil, fmt.Errorf("cannot find prompt for UID %d with the given ID: %s", user, id)
+	}
+	prompt, ok := userEntry.ByID[id]
+	if !ok {
+		return nil, fmt.Errorf("cannot find prompt for UID %d with the given ID: %s", user, id)
+	}
+	allow, err := outcome.IsAllow()
+	if err != nil {
+		return nil, err
+	}
+	for _, listenerReq := range prompt.listenerReqs {
+		if err := sendReply(listenerReq, allow); err != nil {
+			return nil, err
+		}
+	}
+	delete(userEntry.ByID, id)
+	pdb.notifyPrompt(user, id)
+	return prompt, nil
+}
+
+var sendReply = func(listenerReq *listener.Request, reply interface{}) error {
+	return listenerReq.Reply(reply)
+}
+
+// HandleNewRule checks if any existing prompts are satisfied by the given rule
+// contents and, if so, sends back a decision to their listener requests.
+//
+// A prompt is satisfied by the given rule contents if the user, snap,
+// interface, and path of the prompt match those of the rule, and if either the
+// outcome is "allow" and all of the prompt's permissions are matched by those
+// of the rule contents, or if the outcome is "deny" and any of the permissions
+// match.
+//
+// Records a notice for any prompt which was satisfied, or which had some of
+// its permissions satisfied by the rule contents. In the future, only the
+// remaining unsatisfied permissions of a partially-satisfied prompt must be
+// satisfied for the prompt as a whole to be satisfied.
+//
+// Returns the IDs of any prompts which were fully satisfied by the given rule
+// contents.
+func (pdb *PromptDB) HandleNewRule(user uint32, snap string, iface string, constraints *prompting.Constraints, outcome prompting.OutcomeType) ([]string, error) {
+	pdb.mutex.Lock()
+	defer pdb.mutex.Unlock()
+	allow, err := outcome.IsAllow()
+	if err != nil {
+		return nil, err
+	}
+	satisfiedPromptIDs := []string{}
+	userEntry, ok := pdb.perUser[user]
+	if !ok {
+		return satisfiedPromptIDs, nil
+	}
+	for id, prompt := range userEntry.ByID {
+		if !(prompt.Snap == snap && prompt.Interface == iface) {
+			continue
+		}
+		matched, err := constraints.Match(prompt.Constraints.Path)
+		if err != nil {
+			return nil, err
+		}
+		if !matched {
+			continue
+		}
+		modified := prompt.Constraints.subtractPermissions(constraints.Permissions)
+		if !modified {
+			continue
+		}
+		pdb.notifyPrompt(user, id)
+		if len(prompt.Constraints.Permissions) > 0 && allow == true {
+			continue
+		}
+		// All permissions of prompt satisfied, or any permission denied
+		for _, listenerReq := range prompt.listenerReqs {
+			sendReply(listenerReq, allow)
+		}
+		delete(userEntry.ByID, id)
+		satisfiedPromptIDs = append(satisfiedPromptIDs, id)
+	}
+	return satisfiedPromptIDs, nil
+}
+
+// CleanUp removes all outstanding prompts and records a notice for each one.
+//
+// This should be called when snapd is shutting down, to notify prompt clients
```

Comment:
how is that situation distinguished from the client POV from the original notification about the prompt needing reply? Also CleanUp doesn't seem the most clear name for this

Reply:
In short, there's no distinction. Whenever a client receives a prompt notice, they should query the `/v2/interfaces/requests/prompts/{id}` endpoint to get the new state of that prompt (created, modified, replied/cancelled).

The reasoning behind this is we didn't want notices to hold any additional data, the worry being missed signals when a new notice overwrites a previous one which had important data.

But, since prompts can only be replied to/cancelled once, and since the requestprompts lock is held while doing so (and the state lock is held while recording the notice, and the requestprompts lock is not released until that succeeds), I think this is a situation where we could safely include some additional data about the prompt being resolved.

This is something I've been wanting to add for a while (I have an old task about it: https://warthogs.atlassian.net/browse/SNAPDENG-15165), so I'll go ahead and do so, but if you'd prefer not to include data I can revert that change.

Reply:
Regarding the name, I'll switch to `Close()` to match that of the listener and managers in ifacestate.

Reply:
from discussion we are having, it seems it might be that when snapd restarts it will reemit the prompts, do we really want them to be marked cancelled in the mean time though?

Reply:
It will take a fair amount of change in the kernel and listener, and also here, before we are able to persist prompts across snapd restarts, so I don't think it should be a blocker. I think there's nothing wrong with `"resolved": "cancelled"` being included for now, and if in the future prompts aren't cancelled like that anymore, clients will never see it again but I don't see how it would break compatibility.

For a future PR, when we're adding handling for if there's no prompting client connected, we may want to use `"resolved": "cancelled"` for prompts which timeout and are auto-denied after some amount of time where there's no client connected. Or perhaps we'll want to use `"resolved": "timed out"`.

Anyway, I think the more important part is that there's a data field with `"resolved"` as the key, indicating that there's no need to go look up prompt details. And this is just for convenience anyway, AFAIK the aa-prompting-test client looks up prompts details always.

Reply:
it would be good to leave a TODO here about this open point